src/client/serialize/hdf_serialize.c

```diff
@@ -0,0 +1,1034 @@
+		goto out;
+	}
+
+	entry = &prop_query->dpp_entries[11];
+	rc = cont_serialize_prop_str(file_id, entry, "DAOS_PROP_CO_OWNER");
+	if (rc != 0) {
+		goto out;
+	}
+
+	entry = &prop_query->dpp_entries[12];
+	rc = cont_serialize_prop_str(file_id, entry,
+				     "DAOS_PROP_CO_OWNER_GROUP");
+	if (rc != 0) {
+		goto out;
+	}
+
+	entry = &prop_query->dpp_entries[13];
+	rc = cont_serialize_prop_uint(file_id, entry, "DAOS_PROP_CO_DEDUP");
+	if (rc != 0) {
+		goto out;
+	}
+
+	entry = &prop_query->dpp_entries[14];
+	rc = cont_serialize_prop_uint(file_id, entry,
+				      "DAOS_PROP_CO_DEDUP_THRESHOLD");
+	if (rc != 0) {
+		goto out;
+	}
+
+	entry = &prop_query->dpp_entries[15];
+	rc = cont_serialize_prop_uint(file_id, entry,
+				      "DAOS_PROP_CO_ALLOCED_OID");
+	if (rc != 0) {
+		goto out;
+	}
+
+	/* serialize ACL */
+	if (prop_query->dpp_nr > 16) {
+		entry = &prop_query->dpp_entries[16];
+		rc = cont_serialize_prop_acl(file_id, entry,
+					     "DAOS_PROP_CO_ACL");
+		if (rc != 0) {
+			goto out;
+		}
+	}
+
+out:
+	return rc;
+}
+
+int
+cont_serialize_usr_attrs(hid_t *file_id, hid_t *usr_attr_memtype, int num_attrs,
+			 char **names, char **buffers, size_t *sizes,
+			 daos_handle_t coh)
+{
+	int		rc = 0;
+	hid_t		status = 0;
+	hid_t		dset = 0;
+	hid_t		dspace = 0;
+	hsize_t		dims[1];
+	struct usr_attr	*attr_data = NULL;
+	int		i;
+
+	if (num_attrs == 0) {
+		goto out_no_attrs;
+	}
+
+	/* Create the user attribute data space */
+	dims[0] = num_attrs;
+	dspace = H5Screate_simple(1, dims, NULL);
+	if (dspace < 0) {
+		fprintf(stderr, "failed to create user attr dspace");
+		rc = 1;
+		goto out;
+	}
+
+	/* Create the user attribute dataset */
+	dset = H5Dcreate(*file_id, "User Attributes",
+			 *usr_attr_memtype, dspace,
+			 H5P_DEFAULT, H5P_DEFAULT, H5P_DEFAULT);
+	if (dset < 0) {
+		fprintf(stderr, "failed to create user attribute dset");
+		rc = 1;
+		goto out;
+	}
+
+	/* Allocate space for all attributes */
+	attr_data = calloc(num_attrs, sizeof(struct usr_attr));
+	if (attr_data == NULL) {
+		fprintf(stderr, "failed to allocate user attributes");
+		rc = 1;
+		goto out;
+	}
+
+	/* Set the data for all attributes */
+	for (i = 0; i < num_attrs; i++) {
+		attr_data[i].attr_name = names[i];
+		attr_data[i].attr_val.p = buffers[i];
+		attr_data[i].attr_val.len = sizes[i];
+	}
+
+	status = H5Dwrite(dset, *usr_attr_memtype, H5S_ALL,
+			  H5S_ALL, H5P_DEFAULT, attr_data);
+	if (status < 0) {
+		fprintf(stderr, "failed to write user attr dset");
+		rc = 1;
+		goto out;
+	}
+
+out:
+	free(attr_data);
+	H5Dclose(dset);
+	H5Tclose(*usr_attr_memtype);
+	H5Sclose(dspace);
+out_no_attrs:
+	return rc;
+}
+
+int
+serialize_daos_metadata(char *filename, daos_prop_t *props,
+			daos_handle_t coh, int num_attrs,
+			char **names, char **buffers, size_t *sizes)
+{
+	int	rc = 0;
+	hid_t	status;
+	hid_t	file_id;
+	hid_t	usr_attr_memtype;
+	hid_t	usr_attr_name_vtype;
+	hid_t	usr_attr_val_vtype;
+
+	fprintf(stderr, "Writing metadata file: %s\n", filename);
```

Comment:
I'm not sure how messages like this should be printed for this API. We might need to use the standard INFO, DEBUG, ERROR masks.